scripts/install_whonix_qubes/scripts/1-TemplateVM/1.0-haveno-templatevm.sh

```diff
@@ -0,0 +1,189 @@
+	
+	
+	## Install wget
+	apt install -y wget
+	
+	
+	## Function to print messages in blue:
+	echo_blue() {
+	  echo -e "\033[1;34m$1\033[0m"
+	}
+	
+	
+	# Function to print error messages in red:
+	echo_red() {
+	  echo -e "\033[0;31m$1\033[0m"
+	}
+	
+	
+	## Sweep for old release files
+	rm *.asc desktop-*-SNAPSHOT-all.jar.SHA-256 haveno*
+	
+	
+	## Define URL & PGP Fingerprint etc. vars:
+	user_url=$PRECOMPILED_URL
+	base_url=$(printf ${user_url} | awk -F'/' -v OFS='/' '{$NF=""}1')
+	expected_fingerprint=$FINGERPRINT
+	binary_filename=$(awk -F'/' '{ print $NF }' <<< "$user_url")
+	package_filename="haveno.deb"
+	signature_filename="${binary_filename}.sig"
+	key_filename="$(printf "$expected_fingerprint" | tr -d ' ' | sed -E 's/.*(................)/\1/' )".asc
+	wget_flags="--tries=10 --timeout=10 --waitretry=5 --retry-connrefused --show-progress"
+	
+	
+	## Debug:
+	printf "\nUser URL=$user_url\n"
+	printf "\nBase URL=$base_url\n"
+	printf "\nFingerprint=$expected_fingerprint\n"
+	printf "\nBinary Name=$binary_filename\n"
+	printf "\nPackage Name=$package_filename\n"
+	printf "\nSig Filename=$signature_filename\n"
+	printf "\nKey Filename=$key_filename\n"
+	
+	
+	## Configure for tinyproxy:
+	export https_proxy=http://127.0.0.1:8082
+	
+	
+	## Download Haveno binary:
+	echo_blue "Downloading Haveno from URL provided ..."
+	wget "${wget_flags}" -cq "${user_url}" || { echo_red "Failed to download Haveno binary."; exit 1; }
+	
+	
+	## Download Haveno signature file:
+	echo_blue "Downloading Haveno signature ..."
+	wget "${wget_flags}" -cq "${base_url}""${signature_filename}" || { echo_red "Failed to download Haveno signature."; exit 1; }
+	
+	
+	## Download the GPG key:
+	echo_blue "Downloading signing GPG key ..."
+	wget "${wget_flags}" -cqO "${key_filename}" "https://keyserver.ubuntu.com/pks/lookup?op=get&search=0x$(echo "$expected_fingerprint" | tr -d ' ')" || { echo_red "Failed to download GPG key."; exit 1; }
+	
+	
+	## Import the GPG key:
+	echo_blue "Importing the GPG key ..."
+	gpg --import "${key_filename}" || { echo_red "Failed to import GPG key."; exit 1; }
+	
+	
+	## Extract imported fingerprints:
+	imported_fingerprints=$(gpg --with-colons --fingerprint | grep -A 1 'pub' | grep 'fpr' | cut -d: -f10 | tr -d '\n')
+	
+	
+	## Remove spaces from the expected fingerprint for comparison:
+	formatted_expected_fingerprint=$(echo "${expected_fingerprint}" | tr -d ' ')
+	
+	
+	## Check if the expected fingerprint is in the list of imported fingerprints:
+	if [[ ! "${imported_fingerprints}" =~ "${formatted_expected_fingerprint}" ]]; then
+	  echo_red "The imported GPG key fingerprint does not match the expected fingerprint."
+	  exit 1
+	fi
+	
+	
+	## Verify the downloaded binary with the signature:
+	echo_blue "Verifying the signature of the downloaded file ..."
+	OUTPUT=$(gpg --digest-algo SHA256 --verify "${signature_filename}" "${binary_filename}" 2>&1)
+	
+	if ! echo "$OUTPUT" | grep -q "Good signature from"; then
```

Comment:
This hardcoded string should not be searched for.

Instead it should check for GPG's exit code, e.g.

```
if ! gpg --digest-algo SHA256 --verify "${signature_filename}" "${binary_filename}" >/dev/null 2>&1; then
    echo "Signature verification failed!"
    exit 1
fi
```